### DISCUSSION TITLE
[Feature]: Add to album from album

### DISCUSSION THREAD

**Discussion by raisinbear** [immich-app/immich#1617] [Feature Request]
### Feature detail

Thanks again so much for immich! :) I'm really enjoying this project a lot!

I am sorry if this feature has already been wished for, but I can't seem to find it in the open items.

Surprisingly often, I find myself wanting to add at least part of the photos in an album to another album. This may be partly due to my sorting habits, but I think it generally makes sense for shared albums. For instance: You're on a holiday with other people. You add all the holiday pics to an album. Some of those images you want to share with others, some you don't. But from the already sorted album it is so much easier to select the images than from the timeline which in my case is cluttered with other things (screenshots, photos from work someone sent me and so on).

Thanks for considering!

### Platform

Server

**Comment by alextran1502** [comment]
I believe this might be the duplicated of #1010

**Reply by zeyadtamimi** [reply]
Hey Alex, 

From briefly reading #1010 I dont think they are DUPs. They may share some implementation details but fundamentally I think they are different.

Presently you can add a photo to an album by first selecting an image from the main "photos" page, clicking the "+" menu button on the top write and clicking "Add to album".  You can use this flow multiple times to add the same photo to multiple albums.

However, if you are in "Album" page and you select a photo, the "+" menu does not appear and so you cannot add the photo to another album this way.

I believe @raisinbear's request is basically enabling the "+" menu from the album view to facilitate adding images to multiple albums.

**Reply by raisinbear** [reply]
> I believe @raisinbear's request is basically enabling the "+" menu from the album view to facilitate adding images to multiple albums.

Exactly! 😊

**Reply by anddon** [reply]
Hello!
First of all, I just want to say thank you for the work done on the project, I think you have managed to hit a very nice sweet spot in terms of supported features so far and I think the future looks very promising.

Coming back to the feature at hand, IMO what @raisinbear is proposing is a sensible (and pretty much already implemented) 
solution for managing libraries with photos that can belong in multiple albums.

On the lower level, this is already supported (photos CAN already be part of multiple albums) and the only thing that needs done is filling in some small gaps: 
* adding the "+" action in the photo selection context menu in the album view (right not it's only present in the timeline view) (requirement)
* improving on the multiple selection UX (having a select all action button or allowing to select with shift/ctrl etc) (enhancement)

This, IMHO, is a very flexible and generic solution that can cover a lot of use-cases and all this without implementing too much special/specific/complex behaviour or semantics (folders/groups, parents, relationships etc)
A user can:
* go to an existing album
* select all (or just some of) the photos
* click on "+" > "Add to album" 
* select / create a "parent" album from the modal
(identical to what is already possible from the timeline view)

As opposed to the "hardcoded inheritance" model of the album nesting / folders, this more generic solution has the advantage of allowing (but not forcing) the user to only select a subset of photos that he wants to be part of the "parent" / bigger album.
And actually, it can tailor to any scheme of album grouping and naming (in the end, albums are just groups of pointers to the stored assets) and gives the user the freedom to organize as he wishes.

I'm willing to put in the effort and help to come up with a PR to tackle this.

WDYT?

**Reply by raisinbear** [reply]
Thank you very much @anddon for considering. This is precisely what I meant and more. And if it comes to realizing this feature, my sincere thanks to you for the effort! 🙂

**Reply by anddon** [reply]
Likewise, thanks for the feature request. 
While I am willing to help make this happend, I think that before starting to code anything, we need to ensure that this is something that the stewards of the project are on board with. cc @alextran1502 

When it comes to how to actually make it happen, it could either be done by copying the "Add" menu and handlers from the timeline viewer to the album viewer (like you have done), but maybe it's worth extracting the "asset action bar" into a shared/reusable component and have a more uniform set of selection actions, regardless of context.

From what I can see, these are the asset actions exposed so far:

* Share
* Download
* Add
    * Add to Favorites (unless on the favorites view)
    * Add to Album
    * Add to Shared Album
* Delete (which, based on context will either remove from the album/favorites or delete asset entirely)
* Set as Cover (if in an album context and the selection contains a single asset; not implemented like this so far, but it could be an option)

Not sure if there is something I'm missing here and if there are other implications.
Let's maybe wait for some input from some of the contributors and see if/how we can move this forward.

**Reply by alextran1502** [reply]
@anddon I think your proposal is fine. It adds value to the album viewer with already implemented mechanisms. 

And of course, we have to configure the view of what assets are already in the album in the add-to-album selection view. Please feel free to open a PR, and we can onboard this feature

**Reply by anddon** [reply]
Ok, cool. I'll try it out and see how it goes. 
I'm not that versed with Svelte, so it won't be very fast and maybe there will be some back on forth on the PR, until I get the hang of it.

**Comment by raisinbear** [comment]
Hm, this would probably implement it, true. I actually didn’t think quite so sophisticated 😅 Just meant another set of UI elements to directly add photos selected in one album to another (independent) album. Instead of going through the timeline again. Right now I only see the option to remove an image from the current album, not share it to another.
But if the nested thing might happen, all the better 😀

**Comment by raisinbear** [comment]
@alextran1502 Soo, I tried to understand how this whole thing is working in hopes of being able to implement this change myself. However, I quickly found out that this exceeds my capabilities. I'm afraid I doubt I will be able to contribute anything that could be considered quality code any time soon 😅. Making some wild guesses and doing some (dirty) copy pasta I _could_ get it to work the way I intended. The changes below allow me to add the selected photos from one album to new or existing albums and favorites, which is exactly what I wanted to be able to do (minus the favorites part, but that one might be handy as well).
What I couldn't figure out is how to keep the backUrl situation under control. Once a photo is added to a new album, one is redirected to the latter, which makes sense. The backUrl now points to the album one comes from, due to the assignment in afterNavigation(). Clicking the back button leads back to the origin album, the backUrl of which now points to the new album, so endless circle 😑
Anyways, I'll leave this here just to clarify what my initial post was aiming at. I'd be glad if this helps someone with proper knowledge to implement more quickly. In case it is a feature of interest at all, that is.

```diff
   immich-microservices:
     container_name: immich_microservices
diff --git a/web/src/lib/components/album-page/album-viewer.svelte b/web/src/lib/components/album-page/album-viewer.svelte
index aa0f7b42..09011470 100644
--- a/web/src/lib/components/album-page/album-viewer.svelte
+++ b/web/src/lib/components/album-page/album-viewer.svelte
@@ -28,6 +28,8 @@
 	import ThumbnailSelection from './thumbnail-selection.svelte';
 	import ControlAppBar from '../shared-components/control-app-bar.svelte';
 	import CloudDownloadOutline from 'svelte-material-icons/CloudDownloadOutline.svelte';
+	import { assetStore } from '$lib/stores/assets.store';
+	import AlbumSelectionModal from '../shared-components/album-selection-modal.svelte';

 	import {
 		notificationController,
@@ -38,7 +40,7 @@
 	import CreateSharedLinkModal from '../shared-components/create-share-link-modal/create-shared-link-modal.svelte';
 	import ThemeButton from '../shared-components/theme-button.svelte';
 	import { openFileUploadDialog } from '$lib/utils/file-uploader';
-	import { bulkDownload } from '$lib/utils/asset-utils';
+	import { addAssetsToAlbum, bulkDownload } from '$lib/utils/asset-utils';
 	import GalleryViewer from '../shared-components/gallery-viewer/gallery-viewer.svelte';
 	import ImmichLogo from '../shared-components/immich-logo.svelte';

@@ -61,10 +63,13 @@
 			}
 		}
 	}
+	let addToSharedAlbum = false;
 	let isShowShareUserSelection = false;
 	let isEditingTitle = false;
 	let isCreatingSharedAlbum = false;
 	let isShowShareInfoModal = false;
+	let isShowAddMenu = false;
+	let isShowAlbumPicker = false;
 	let isShowAlbumOptions = false;
 	let isShowThumbnailSelection = false;

@@ -210,6 +215,72 @@
 		}
 	};

+	const handleShowMenu = ({ x, y }: MouseEvent) => {
+		contextMenuPosition = { x, y };
+		isShowAddMenu = !isShowAddMenu;
+	};
+
+	const handleAddToFavorites = () => {
+		isShowAddMenu = false;
+
+		let cnt = 0;
+		for (const asset of multiSelectAsset) {
+			if (!asset.isFavorite) {
+				api.assetApi.updateAsset(asset.id, {
+					isFavorite: true
+				});
+				assetStore.updateAsset(asset.id, true);
+				cnt = cnt + 1;
+			}
+		}
+
+		notificationController.show({
+			message: `Added ${cnt} to favorites`,
+			type: NotificationType.Info
+		});
+
+		isMultiSelectionMode = false;
+		clearMultiSelectAssetAssetHandler();
+	};
+
+	const handleShowAlbumPicker = (shared: boolean) => {
+		isShowAddMenu = false;
+		isShowAlbumPicker = true;
+		addToSharedAlbum = shared;
+	};
+
+	const handleAddToNewAlbum = (event: CustomEvent) => {
+		isShowAlbumPicker = false;
+
+		const { albumName }: { albumName: string } = event.detail;
+		const assetIds = Array.from(multiSelectAsset).map((a) => a.id);
+		api.albumApi.createAlbum({ albumName, assetIds }).then((response) => {
+			const { id, albumName } = response.data;
+
+			notificationController.show({
+				message: `Added ${assetIds.length} to ${albumName}`,
+				type: NotificationType.Info
+			});
+
+			isMultiSelectionMode = false;
+			clearMultiSelectAssetAssetHandler();
+
+			goto('/albums/' + id);
+		});
+	};
+
+	const handleAddToAlbum = async (event: CustomEvent<{ album: AlbumResponseDto }>) => {
+		isShowAlbumPicker = false;
+		const album = event.detail.album;
+
+		const assetIds = Array.from(multiSelectAsset).map((a) => a.id);
+
+		addAssetsToAlbum(album.id, assetIds).then(() => {
+			isMultiSelectionMode = false;
+			clearMultiSelectAssetAssetHandler();
+		});
+	};
+
 	const sharedUserDeletedHandler = async (event: CustomEvent) => {
 		const { userId }: { userId: string } = event.detail;

@@ -389,6 +460,7 @@
 					on:click={handleDownloadSelectedAssets}
 					logo={CloudDownloadOutline}
 				/>
+				<CircleIconButton title="Add" logo={Plus} on:click={handleShowMenu} />
 				{#if isOwned}
 					<CircleIconButton
 						title="Remove from album"
@@ -549,6 +621,26 @@
 	</section>
 </section>

+{#if isShowAddMenu}
+	<ContextMenu {...contextMenuPosition} on:clickoutside={() => (isShowAddMenu = false)}>
+		<div class="flex flex-col rounded-lg ">
+			<MenuOption on:click={handleAddToFavorites} text="Add to Favorites" />
+			<MenuOption on:click={() => handleShowAlbumPicker(false)} text="Add to Album" />
+			<MenuOption on:click={() => handleShowAlbumPicker(true)} text="Add to Shared Album" />
+		</div>
+	</ContextMenu>
+{/if}
+
+{#if isShowAlbumPicker}
+	<AlbumSelectionModal
+		shared={addToSharedAlbum}
+		on:newAlbum={handleAddToNewAlbum}
+		on:newSharedAlbum={handleAddToNewAlbum}
+		on:album={handleAddToAlbum}
+		on:close={() => (isShowAlbumPicker = false)}
+	/>
+{/if}
+
 {#if isShowAssetSelection}
 	<AssetSelection
 		albumId={album.id}
```

**Comment by jrasm91** [comment]
This was implemented a few releases ago on the web

**Reply by RemDelaporteMathurin** [reply]
Has it? I can't find it on the web app (on my phone). I can add individual pictures from an album to another one but not a set
